dataprofiler/profilers/profile_builder.py

```diff
@@ -1402,17 +1419,46 @@ def tqdm(l):
         # Newly introduced features (python3.8) improves the situation
         sample_ids = np.array(sample_ids)
 
-        # Create structured profile objects
-        new_cols = set()
-        for col in data.columns:
-            if col not in self._profile:
-                self._profile[col] = StructuredColProfiler(
-                    sample_size=sample_size,
-                    min_true_samples=min_true_samples,
-                    sample_ids=sample_ids,
-                    options=self.options
-                )
-                new_cols.add(col)
+        # Create StructuredColProfilers (must be either first initialization
+        # or unique column names)
+        new_cols = False
+        if not duplicate_cols_present:
+            # Given duplicate columns despite duplicate columns not being
+            # present in nonempty _profile
+            if initialized and duplicate_cols_given:
+                raise ValueError("Attempted to update data with duplicate "
+                                 "column names that weren't present before "
+                                 "update. Schema must be identical when "
+                                 "profiling data with duplicate column names.")
+
+            # Either initializing _profile for the first time or updating
+            # _profile that doesn't contain duplicate column names
+            for col in data.columns:
+                # If initializing for the first time, must fill mapping
+                # If initialized, with no duplicate columns, only add to mapping
+                # if not already there, since this means we are updating
+                if not initialized or col not in self._col_name_to_idx:
+                    # Append StructuredColProfiler to list of profiles
+                    # and record index where it was appended to in list
+                    self._col_name_to_idx.setdefault(col, []).append(
+                        len(self._profile))
+                    self._profile.append(StructuredColProfiler(
+                        sample_size=sample_size,
+                        min_true_samples=min_true_samples,
+                        sample_ids=sample_ids,
+                        options=self.options
+                    ))
+                    new_cols = True
+        else:
+            # Ensure same schema if initialized with duplicate columns
+            # No adding columns to _profile after initialization in this case
+            mapping_given = dict()
+            for i in range(len(data.columns)):
+                col = data.columns[i]
+                mapping_given.setdefault(col, []).append(i)
+            if mapping_given != self._col_name_to_idx:
+                raise ValueError("Schema of data with duplicate column names "
+                                 "not respected when updating profile.")
```

Comment:
this should be checked above prior to any calculation. 

However, is it no adding to columns period or no adding to a schema which has duplicate keys and you are adding the same key name?

Reply:
The logic is as follows:
If unitinitialized:
- Add a blank SCP for each column, and add each column to mapping (can be duplicate)

If initialized:
- If no duplicate cols in initial data, can add columns that weren't previously present, can update columns already present, but cannot give duplicate columns in data given to update
- If duplicate cols in initial data, no adding of cols at update, update data must look identical to data given at initialization